src/Mod/Sketcher/App/SketchObject.cpp

```diff
@@ -3644,113 +3649,85 @@ int SketchObject::trim(int GeoId, const Base::Vector3d& point)
+    newConstr->Second = cuttingGeoId;
+    if (isPointAtPosition(obj, cuttingGeoId, PointPos::start, cutPointVec)) {
+        newConstr->Type = Sketcher::Coincident;
+        newConstr->SecondPos = PointPos::start;
+    }
+    else if (isPointAtPosition(obj, cuttingGeoId, PointPos::end, cutPointVec)) {
+        newConstr->Type = Sketcher::Coincident;
+        newConstr->SecondPos = PointPos::end;
+    }
+    else {
+        // Points are sufficiently far apart: use point-on-object
+        newConstr->Type = Sketcher::PointOnObject;
+        newConstr->SecondPos = PointPos::none;
+    }
+    return newConstr;
+};
 
-    //******************* Step A => Detection of intersection - Common to all Geometries
-    //****************************************//
-    // GeoIds intersecting the curve around `point`
-    std::array<int, 2> cuttingGeoIds {GeoEnum::GeoUndef, GeoEnum::GeoUndef};
-    // Points at the intersection
-    std::array<Base::Vector3d, 2> cutPoints;
+bool isGeoIdAllowedForTrim(const SketchObject* obj, int GeoId)
+{
+    const auto* geo = obj->getGeometry(GeoId);
 
-    // Using SketchObject wrapper, as Part2DObject version returns GeoId = -1 when intersection not
-    // found, which is wrong for a GeoId (axis). seekTrimPoints returns:
-    // - For a parameter associated with "point" between an intersection and the end point
-    // (non-periodic case) cuttingGeoIds[0] != GeoUndef and cuttingGeoIds[1] == GeoUndef
-    // - For a parameter associated with "point" between the start point and an intersection
-    // (non-periodic case) cuttingGeoIds[1] != GeoUndef and cuttingGeoIds[0] == GeoUndef
-    // - For a parameter associated with "point" between two intersection points, cuttingGeoIds[0]
-    // != GeoUndef and cuttingGeoIds[1] != GeoUndef
-    //
-    // FirstParam < point1param < point2param < LastParam
-    if (!SketchObject::seekTrimPoints(GeoId,
-                                      point,
-                                      cuttingGeoIds[0],
-                                      cutPoints[0],
-                                      cuttingGeoIds[1],
-                                      cutPoints[1])) {
-        // If no suitable trim points are found, then trim defaults to deleting the geometry
-        delGeometry(GeoId);
-        return 0;
-    }
+    return GeoId >= 0 && GeoId <= obj->getHighestCurveIndex()
+        && GeometryFacade::isInternalType(geo, InternalType::None);
+}
 
-    const auto* geoAsCurve = static_cast<const Part::GeomCurve*>(geo);
+bool getParamLimitsOfNewGeosForTrim(const SketchObject* obj,
+                                    int GeoId,
+                                    std::array<int, 2>& cuttingGeoIds,
+                                    std::array<Base::Vector3d, 2>& cutPoints,
+                                    std::vector<std::pair<double, double>>& paramsOfNewGeos)
+{
+    const auto* geoAsCurve = static_cast<const Part::GeomCurve*>(obj->getGeometry(GeoId));

```

Comment:
```suggestion
const auto* geoAsCurve = obj->getGeometry<Part::GeomCurve>(GeoId);
```

Reply:
Didn't know this templating exists. Not sure what opinion I have about it.

Reply:
I prefer the template function as:
* I find it easier to read
* It adds compile time checks for the type

Reply:
Done locally.